cirq/sim/wave_function.py

```diff
@@ -345,36 +357,137 @@ def to_valid_state_vector(
         qid_shape = (2,) * num_qubits
+    else:
+        qid_shape = tuple(qid_shape)
+    if num_qubits is None:
+        num_qubits = len(qid_shape)
     if num_qubits != len(qid_shape):
         raise ValueError('num_qubits != len(qid_shape). num_qubits is <{!r}>. '
                          'qid_shape is <{!r}>.'.format(num_qubits, qid_shape))
-    if isinstance(state_rep, np.ndarray):
-        if len(state_rep) != np.prod(qid_shape, dtype=int):
-            raise ValueError(
-                'initial state was of size {} '
-                'but expected state for {} qubits with qid shape {}'.format(
-                    len(state_rep), num_qubits, qid_shape))
-        state = state_rep.copy()
-    elif isinstance(state_rep, int):
-        if state_rep < 0:
-            raise ValueError('initial_state must be positive')
-        elif state_rep >= np.prod(qid_shape, dtype=int):
-            raise ValueError(
-                'initial state was {} but expected state for {} qubits'.format(
-                    state_rep, num_qubits))
-        else:
-            state = linalg.one_hot(shape=np.prod(qid_shape, dtype=int),
-                                   dtype=dtype,
-                                   index=state_rep)
-    else:
-        raise TypeError('initial_state was not of type int or ndarray')
-    validate_normalized_state(state,
+
+    tensor = _state_like_to_state_tensor(state_like=state_rep,
+                                         qid_shape=qid_shape,
+                                         dtype=dtype,
+                                         atol=atol)
+    return tensor.reshape(tensor.size)  # Flatten.
+
+
+def _state_like_to_state_tensor(*, state_like: 'cirq.STATE_VECTOR_LIKE',
+                                qid_shape: Tuple[int, ...],
+                                dtype: Type[np.number], atol) -> np.ndarray:
+
+    if isinstance(state_like, int):
+        return _computational_basis_state_to_state_tensor(state=state_like,
+                                                          qid_shape=qid_shape,
+                                                          dtype=dtype)
+
+    if isinstance(state_like, Sequence):
+        converted = np.array(state_like)
+        if converted.shape:
+            state_like = converted
+    if isinstance(state_like, np.ndarray):
+        prod = np.prod(qid_shape, dtype=int)
+
+        if len(qid_shape) == prod:
+            if (not isinstance(state_like, np.ndarray) or
+                    state_like.dtype.kind != 'c'):
+                raise ValueError(
+                    'Because len(qid_shape) == product(qid_shape), it is '
+                    'ambiguous whether or not the given `state_like` is a '
+                    'state vector or a list of computational basis values for '
+                    'the qudits. In this situation you are required to pass '
+                    'in a state vector that is a numpy array with a complex '
+                    'dtype.')
```

Comment:
This is worrying.  As a user, I would it expect it to work for all input sizes if it works for most input sizes.

Reply:
Because `2**n > n` for all `n >= 0`, you never see this unless you have qudits with dimension 1. I think this is a strange enough corner case that we don't have to ensure it works perfectly.

We could reduce the number of relevant cases by checking if the array has non-integer entries in it (in which case it must be a state vector) or if it has integers summing to more than one (in which case it must be a qudit value list). The only truly ambiguous case is something like `shape=(2, 2, 1, 1), value=[1, 0, 0, 0]` where the value is one-hot and has an ambiguous length.

Reply:
tfw your qudits have dimension 1